server/src/main/java/org/elasticsearch/index/query/AutomatonQueryBuilder.java

```
@@ -0,0 +1,85 @@
+
+/**
+ * Implements an Automaton query, which matches documents based on a Lucene Automaton.
+ * It does not support serialization or XContent representation,
+ */
+public class AutomatonQueryBuilder extends AbstractQueryBuilder<AutomatonQueryBuilder> implements MultiTermQueryBuilder {
+    private final String fieldName;
+    private final Automaton automaton;
+
+    public AutomatonQueryBuilder(String fieldName, Automaton automaton) {
+        if (Strings.isEmpty(fieldName)) {
+            throw new IllegalArgumentException("field name is null or empty");
+        }
+        if (automaton == null) {
+            throw new IllegalArgumentException("automaton cannot be null");
+        }
+        this.fieldName = fieldName;
+        this.automaton = automaton;
+    }
+
+    @Override
+    public String fieldName() {
+        return fieldName;
+    }
+
+    @Override
+    public String getWriteableName() {
+        throw new UnsupportedOperationException("AutomatonQueryBuilder does not support getWriteableName");
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        throw new UnsupportedEncodingException("AutomatonQueryBuilder does not support doWriteTo");
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        throw new UnsupportedEncodingException("AutomatonQueryBuilder does not support doXContent");
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        return new AutomatonQuery(new Term(fieldName), automaton);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(fieldName, automaton);
+    }
+
+    @Override
+    protected boolean doEquals(AutomatonQueryBuilder other) {
+        return Objects.equals(fieldName, other.fieldName) && Objects.equals(automaton, other.automaton);
+    }
+
+    // TO DO, what should be the minimal supported version?
```

We don't serialize at all so I'd `throw new UnsupportedOperationException`

---

Done